src/main/java/io/dockstore/provision/GSPlugin.java

```diff
@@ -0,0 +1,271 @@
+                System.err.println("Could not create destination path. IO exception:" + e.getMessage());
+                return false;
+            }
+
+            ProgressPrinter printer = new ProgressPrinter();
+
+            FileOutputStream fileContent;
+            try {
+                fileContent = new FileOutputStream((destination.toFile()));
+            } catch (FileNotFoundException e) {
+                System.err.println("File not found exception:" + e.getMessage());
+                return false;
+            }
+
+            PrintStream writeTo = new PrintStream(fileContent);
+
+            if (blob.getSize() < MAX_FILE_SIZE_FOR_SINGLE_WRITE) {
+                // Blob is small read all its content in one request
+                byte[] content = blob.getContent();
+                try {
+                    writeTo.write(content);
+                } catch (IOException e) {
+                    System.err.println("Could not write download file. IO exception:" + e.getMessage());
+                    return false;
+                }
+            } else {
+                // When Blob size is big or unknown use the blob's channel reader.
+                try (ReadChannel reader = blob.reader()) {
+                    WritableByteChannel channel = Channels.newChannel(writeTo);
+                    ByteBuffer bytes = ByteBuffer.allocate(DOWNLOAD_BUFFER_SIZE);
+                    try {
+                        int limit;
+                        long runningTotal = 0;
+                        while ((limit = reader.read(bytes)) > 0) {
+                            bytes.flip();
+                            channel.write(bytes);
+                            bytes.clear();
+                            runningTotal = runningTotal + limit;
+                            printer.handleProgress(runningTotal, inputSize);
+                        }
+                    } catch (IOException e) {
+                        System.err.println("Could not download file. IO exception:" + e.getMessage());
+                        return false;
+                    }
+
+                }
+            }
+
+            writeTo.close();
+            return true;
+        }
+
+        public boolean uploadTo(String destPath, Path sourceFile, Optional<String> metadata) {
+            long inputSize = sourceFile.toFile().length();
+
+            String contentType = "application/octet-stream";
+            try {
+                contentType = Files.probeContentType(sourceFile);
+            } catch (IOException e) {
+                System.err.println("Could not get content type of file. Using default content type. IO exception:" + e.getMessage());
+            }
+
+            // Bug on Mac where Files.probeContentType returns null
+            if (contentType == null)
+                contentType = "application/octet-stream";
+
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = destPath.replace("gs://", "");
```

Comment:
Would put this in a unit tested function.

Reply:
Refactored to separate method and test added